src/analyses/ai.h

```diff
@@ -254,92 +255,133 @@ class ai_baset
   bool visit_edge(
-    locationt l,
+    const tracet &h,
     working_sett &working_set,
     const locationt &to_l,
     const namespacet &ns);
 
   // function calls
+
   bool do_function_call_rec(
-    locationt l_call, locationt l_return,
+    const tracet &h_call,
     const exprt &function,
     const exprt::operandst &arguments,
     const goto_functionst &goto_functions,
     const namespacet &ns);
 
   bool do_function_call(
-    locationt l_call, locationt l_return,
+    const tracet &h_call,
     const goto_functionst &goto_functions,
     const goto_functionst::function_mapt::const_iterator f_it,
     const exprt::operandst &arguments,
     const namespacet &ns);
 
   // abstract methods
+  // These delegate anything that requires knowing the actual type of
+  // the tracet or statet (as opposed to their parent class / interface).
 
-  virtual bool merge(const statet &src, locationt from, locationt to)=0;
+  virtual bool
+  merge(const statet &src, const tracet &from, const tracet &to) = 0;
   // for concurrent fixedpoint
   virtual bool merge_shared(
     const statet &src,
     locationt from,
     locationt to,
     const namespacet &ns)=0;
-  virtual statet &get_state(locationt l)=0;
-  virtual const statet &find_state(locationt l) const=0;
+  virtual statet &get_state(const tracet &h) = 0;
+  virtual const statet &find_state(const tracet &h) const = 0;
   virtual std::unique_ptr<statet> make_temporary_state(const statet &s)=0;
+
+  virtual history_ptrt start_history(locationt bang) = 0;
+  typedef ai_history_baset::step_returnt step_returnt;
+  virtual step_returnt step(const tracet &t, locationt to_l) = 0;
+  virtual working_sett get_history(locationt &l) = 0;
 };
 
 /// Creation, storage and other operations dependent
 /// on the exact types of abstraction used
+/// historyT is expected to be derived from ai_history_baset (a.k.a. historyt)
 /// domainT is expected to be derived from ai_domain_baset (a.k.a. domaint)
-template <typename domainT>
+template <typename historyT, typename domainT>
 class ai_storaget : public ai_baset
 {
 public:
+  typedef historyT historyt;
   typedef domainT domaint;
 
   // constructor
   ai_storaget() : ai_baset()
   {
   }
 
-  typedef goto_programt::const_targett locationt;
-
   /// Direct access to the state map
   /// Unlike abstract_state_* this requires/has knowledge of the template types
-  domainT &operator[](locationt l)
+  domainT &operator[](const historyT &h)
   {
-    return static_cast<domainT &>(get_state(l));
+    return static_cast<domainT &>(get_state(h));
   }
 
-  const domainT &operator[](locationt l) const
+  const domainT &operator[](const historyT &h) const
   {
-    return static_cast<const domainT &>(find_state(l));
+    return static_cast<const domainT &>(find_state(h));
+  }
+
+  void clear() override
+  {
+    history_map.clear();
+    ai_baset::clear();
+  }
+
+  working_sett get_history(locationt &l) override
```

Comment:
Should the return value be copied, const, or writeable?

Reply:
Point.  It was a copy due to C++ type system weirdness.  Post refactor, it can be const and now it is!